src/main/java/software/amazon/awssdk/crt/CrtResource.java

```diff
@@ -16,96 +16,291 @@
+     *
+     * Some classes don't represent an actual native resource.  Instead, they just want to use
+     * the reference and cleanup framework.  See CrtBufferPool for an example.
+     *
+     * A solution that avoids this complication while not changing the buffer/buffer pool properties dramatically would
+     * be welcomed.
+     */
+
+    /**
+     * Is this an actual native resource (false) or does it just track native resources and use the close/shutdown/referencing
+     * aspects (true)?
+     */
+    protected boolean isAlwaysNullResource() { return false; }
+
+    /**
+     * Decrements the reference count to this resource.  If zero is reached, begins (and possibly completes) the resource's
+     * cleanup process.
+     */
     @Override
     public void close() {
-        if (!isNull()) {
-            throw new IllegalStateException("Closing sub-resources before releasing parent Resource may cause use-after-free bugs!");
+        int remainingRefs = refCount.decrementAndGet();
+
+        if (debugNativeObjects) {
+            Log.log(Log.LogLevel.Trace, String.format("Closing instance of class %s with %d remaining refs", this.getClass().getCanonicalName(), remainingRefs));
+        }
+
+        if (remainingRefs > 0) {
+            return;
         }
 
-        while(ownedSubResources.size() > 0) {
-            CrtResource r = ownedSubResources.pop();
+        release();
+
+        if (canReleaseReferencesImmediately()) {
+            releaseReferences();
+        }
+    }
+
+    /**
+     * Decrements the ref counts for all resources referenced by this resource.  Most resources will have this called
+     * from their close() function, but resources with asynchronous shutdown processes must have it called from a
+     * shutdown completion callback.
+     */
+    protected void releaseReferences() {
+        if (debugNativeObjects) {
+            Log.log(Log.LogLevel.Trace, String.format("Instance of class %s closing referenced objects", this.getClass().getCanonicalName()));
+        }
+
+        while(referencedResources.size() > 0) {
+            CrtResource r = referencedResources.pop();
             r.close();
         }
     }
+
+    /**
+     * Debug method to log all of the currently un-closed CRTResource objects.
+     */
+    public static void logNativeResources() {
+        Log.log(Log.LogLevel.Trace, "Dumping native object set:");
+        for (Map.Entry<Long, String> entry : NATIVE_RESOURCES.entrySet()) {
+            Log.log(Log.LogLevel.Trace, String.format(" * %s class instance using native pointer %d", entry.getValue(), entry.getKey().longValue()));
+        }
+    }
+
+    /**
+     * Debug method to increment the current native object count.
+     */
+    private static void incrementNativeObjectCount() {
+        if (!debugNativeObjects) {
+            return;
+        }
+
+        lock.lock();
+        try {
+            ++resourceCount;
```

Comment:
Is there a reason you're not using `AtomicInteger`? Seems like it would be simpler and less error prone than manually managing a lock and int.

Reply:
Not a Java expert, but Conditions need to be associated with a lock and it's a runtime error to signal a condition without the lock being held so AtomicInteger doesn't seem to buy anything.

Reply:
If an auto-close lock already existed I'd happily use it but I'm reluctant to add it just for this.

Reply:
That's fair.